### PR TITLE
Store the last used directory as a preference (#1604)

### LINKED ISSUE #1604: Add a preference option for keeping the last directory used
Derived from https://github.com/CARTAvis/carta-backend/issues/807:

> Additionally, a common behaviour of desktop apps that users may want to enable instead of a static directory is the last directory used (which would require us to store this state somewhere, per-user).

### PR DESCRIPTION
**Pull request by jolopezl**
Closes #1604 adding a preference option for keeping the last used directory.

preference

### CODE REVIEW

File: src/components/HelpDrawer/HelpContent/PreferencesHelpComponent.tsx [hunk 1]
```diff
@@ -27,6 +27,7 @@ export class PreferencesHelpComponent extends React.Component {
                     <li>WCS matching on append: trigger WCS matching automatically for newly appended images</li>
                     <li>Spectral matching: spectral convention to be used for spectral matching of image cubes</li>
                     <li>Transparent image background: when this is enabled, the exported png image will have a transparent background. When it is disabled (default), a white or a black background is added depending on the GUI theme.</li>
+                    <li>Save last used directory: when this is enabled, the file browser will start in the same folder the next session.</li>
```
Comment (veggiesaurus) [inline src/components/HelpDrawer/HelpContent/PreferencesHelpComponent.tsx:30]: thank you for doing this! we always forget to add these and then @kswang1029 has to sift through all the changes :see_no_evil:

File: src/stores/FileBrowserStore.ts [hunk 2]
```diff
@@ -347,6 +348,26 @@ export class FileBrowserStore {
                 this.startingDirectory = this.fileList.directory;
             }
         }
+        if (PreferenceStore.Instance.keepLastUsedFolder) {
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts]: `const preferences = PreferenceStore.Instance` and reuse that within the function

File: src/stores/FileBrowserStore.ts [hunk 3]
```diff
@@ -347,6 +348,26 @@ export class FileBrowserStore {
...
+            console.debug("coconut... saving last directory: " + this.startingDirectory);
+            console.debug("coconut... PreferenceStore.Instance.lastUsedFolder = " + PreferenceStore.Instance.lastUsedFolder);
+        } else {
+            PreferenceStore.Instance.setPreference(PreferenceKeys.GLOBAL_SAVED_LAST_FOLDER, "");
+        }
+    }
+
+    setStartingDirectory() {
+        console.debug("coconut... setting the starting directory...");
+        if (PreferenceStore.Instance.keepLastUsedFolder) {
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts]: see above

File: src/stores/FileBrowserStore.ts [hunk 4]
```diff
@@ -347,6 +348,26 @@ export class FileBrowserStore {
...
+            console.debug("coconut... PreferenceStore.Instance.lastUsedFolder = " + PreferenceStore.Instance.lastUsedFolder);
+            if (PreferenceStore.Instance.lastUsedFolder.length > 0) {
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts]: use `lastUsedFolder?.length` because it could be undefined

File: src/stores/PreferenceStore.ts [hunk 5]
```diff
@@ -82,7 +84,7 @@ export enum PreferenceKeys {
     CATALOG_TABLE_SEPARATOR_POSITION = "catalogTableSeparatorPosition",
 
     PIXEL_GRID_VISIBLE = "pixelGridVisible",
-    PIXEL_GRID_COLOR = "pixelGridColor"
+    PIXEL_GRID_COLOR = "pixelGridColor",
```
Comment (veggiesaurus) [inline src/stores/PreferenceStore.ts]: make sure your IDE's auto-formatting aligns with our `.prettierrc` file. We don't have trailing commas

Comment (jolopezl) [reply src/stores/PreferenceStore.ts]: all set now, thanks.

File: src/components/HelpDrawer/HelpContent/PreferencesHelpComponent.tsx (hunk 1 above)
Comment (jolopezl) [reply src/components/HelpDrawer/HelpContent/PreferencesHelpComponent.tsx:30]: I was lucky to catch this in the latest changes merged into the codebase! I hope I keep the good practice.

File: src/stores/FileBrowserStore.ts [hunk 6]
```diff
@@ -338,6 +338,8 @@ export class FileBrowserStore {
     }
 
     @action saveStartingDirectory(directory?: string) {
+        const preferenceStore = PreferenceStore.Instance;
+        console.debug("coconut... previous last directory: " + preferenceStore.lastUsedFolder);
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts]: you forgot your coconuts :smile:

File: src/stores/FileBrowserStore.ts [hunk 7]
```diff
@@ -347,6 +349,27 @@ export class FileBrowserStore {
                 this.startingDirectory = this.fileList.directory;
             }
         }
+        if (preferenceStore.keepLastUsedFolder) {
+            preferenceStore.setPreference(PreferenceKeys.GLOBAL_SAVED_LAST_FOLDER, this.startingDirectory);
+            console.debug("coconut... saving last directory: " + this.startingDirectory);
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts]: :coconut:

File: src/stores/FileBrowserStore.ts [hunk 8]
```diff
@@ -347,6 +349,27 @@ export class FileBrowserStore {
...
+            console.debug("coconut... PreferenceStore.Instance.lastUsedFolder = " + preferenceStore.lastUsedFolder);
+        } else {
+            preferenceStore.setPreference(PreferenceKeys.GLOBAL_SAVED_LAST_FOLDER, "");
+        }
+    }
+
+    setStartingDirectory() {
+        const preferenceStore = PreferenceStore.Instance;
+        console.debug("coconut... setting the starting directory...");
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts]: :coconut:

File: src/stores/FileBrowserStore.ts [hunk 9]
```diff
@@ -347,6 +349,27 @@ export class FileBrowserStore {
...
+        if (preferenceStore.keepLastUsedFolder) {
+            console.debug("coconut... PreferenceStore.Instance.lastUsedFolder = " + preferenceStore.lastUsedFolder);
+            if (preferenceStore.lastUsedFolder?.length > 0) {
+                this.startingDirectory = preferenceStore.lastUsedFolder;
+                console.debug("coconut... using last used folder: " + this.startingDirectory);
```
Comment (veggiesaurus) [inline src/stores/FileBrowserStore.ts]: :coconut:

**Review by veggiesaurus — CHANGES_REQUESTED** [review]
debug logs :innocent:

File: src/stores/FileBrowserStore.ts (hunk 6 above)
Comment (jolopezl) [reply src/stores/FileBrowserStore.ts]: next time I will leave `pineapples...`!

**Review by veggiesaurus — APPROVED** [review]

**Comment by kswang1029** [comment]
if the folder is gone or renamed, we will see this. It is probably okay to have this "behavior". We can still click the home button to access top_level_folder. 

Screen Shot 2021-08-10 at 7 02 04 PM

**Pull request merged**